b2sdk/transfer/emerge/planner/planner.py

```diff
@@ -0,0 +1,493 @@
+            if current_intent is None:
+                current_intent = intent
+                current_end = fragment_end
+                continue
+
+            if intent is current_intent:
+                current_end = fragment_end
+            else:
+                current_len = current_end - buff_end
+                buff_len = buff_end - buff_start
+                if current_intent.is_copy() and current_len >= min_part_size:
+                    if buff and buff_len < min_part_size:
+                        missing_length = min_part_size - buff_len
+                    else:
+                        missing_length = 0
+                    if missing_length > 0 and current_len - missing_length < min_part_size:
+                        # current intent is *not* a "small copy", but upload buffer is small
+                        # and current intent is too short to build two parts at least the minimum part size
+                        # so we append current intent to upload buffer
+                        buff.append((current_intent, current_end))
+                    else:
+                        if missing_length > 0:
+                            # we "borrow" a fragment of current intent to upload buffer
+                            # to fill it to minimum part size
+                            buff_end += missing_length
+                            buff.append((current_intent, buff_end))
+                        for sub_buff_start, sub_buff in self._buff_split(buff, buff_start):
+                            yield self._get_upload_part(sub_buff, sub_buff_start)
+                        for part in self._get_copy_parts(current_intent, buff_end, current_end):
+                            yield part
+                        buff = []
+                        buff_start = current_end
+                        buff_end = buff_start
+                else:
+                    buff.append((current_intent, current_end))
+                    buff_end = current_end
+                    buff_parts = list(self._buff_split(buff, buff_start))
+                    for sub_buff_start, sub_buff in buff_parts[:-1]:
+                        yield self._get_upload_part(sub_buff, sub_buff_start)
+                    buff_start, buff = buff_parts[-1]
+                current_intent = intent
+                current_end = fragment_end
+                if current_intent is None:
+                    for sub_buff_start, sub_buff in self._buff_split(buff, buff_start):
+                        yield self._get_upload_part(sub_buff, sub_buff_start)
+
+    def _get_upload_part(self, buff, buff_start):
+        if len(buff) == 1 and buff[0][0].is_upload():
+            intent, buff_end = buff[0]
+            relative_offset = buff_start - intent.destination_offset
+            length = buff_end - buff_start
+            definition = UploadEmergePartDefinition(intent.outbound_source, relative_offset, length)
```

Comment:
return it and fold down

Reply:
reject - it has to be wrapped with `EmergePart` - maybe `EmergePart` interface would change in the future, so I want to wrap it once...